### PR TITLE
Phase 2 tracker: template CPE

### PR DESCRIPTION
**Pull request by cmantill**
#### PR description:

This PR integrates the template method for the pixel hit reconstruction in the Cluster Parameter Estimators (CPEs) for Phase 2. The templates have been produced for 5 different geometries in a "temporary" configuration that has been described in detail here [1]. The PR introduces the following changes:

1.  Provide phase 2 conditions for geometries: T5, T6, T14, T15, T16 with latest DB conditions.
2.  Remove phase 2 customizations forcing CPE generic method.
3. Increase array sizes to allow for larger angle space of the phase 2 templates.

@mmusich @tsusa @OzAmram @tvami

#### PR validation:

Tested with phase 2 workflows: `runTheMatrix.py --what upgrade -l 20007.0,20407.0,20807.0,21607.0,22007.0,22807.0,22834.0  -t 4 -j 8`.

The template method for pixel hit residuals has been also tested running the reconstruction step for 100K events and results will be shown here [2]. Since the templates have been produced under "temporary" conditions, we do not expect a big change in the resolution results obtained with this method.

[1] https://indico.cern.ch/event/838960/contributions/3630543/attachments/1941870/3220114/CMS_phase2tem_Nov5.pdf
[2] https://indico.cern.ch/event/838961/

#### if this PR is a backport please specify the original PR:

NA

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmantill 
please clarify the scope of the "RFC".
Is this for the tracker community?
Is the software already adequate for physics or is this just a prototype which is known to need more work.
Thank you.

**Comment by cmantill** [comment]
@slava77 This is already adequate for physics (or at least more adequate than the dummy templates before). RFC is within the scope of the tracker community and just pending for the comments from tomorrow's tracker simulation meeting.

File: CondFormats/SiPixelTransient/interface/SiPixelTemplate.h [hunk 1]
```diff
@@ -207,12 +207,12 @@ struct SiPixelTemplateStore {  //!< template storage structure
...
   float cotbetaY[60];
-  float cotbetaX[5];
-  float cotalphaX[29];
+  float cotbetaX[60];
+  float cotalphaX[60];
   //!< 60 Barrel y templates spanning cluster lengths from 0px to +18px [28 entries for fpix]
   SiPixelTemplateEntry enty[60];
   //!< 29 Barrel x templates spanning cluster lengths from -6px (-1.125Rad) to +6px (+1.125Rad) in each of 5 slices [3x29 for fpix]
-  SiPixelTemplateEntry entx[5][29];
+  SiPixelTemplateEntry entx[60][60];
```
Comment (makortel) [inline CondFormats/SiPixelTransient/interface/SiPixelTemplate.h]: What is the impact on memory? (my quick estimate would indicate that the `sizeof(SiPixelTemplateStore)` would be something like 7.4 MB, but I could have missed someting). Is an `SiPixelTemplateStore` object stored in stack anywhere? (just wondering if there is any risk to hit stack size issues like in #22458)

**Comment by jfernan2** [comment]
@cmantill  I just want you bring your attention to the large number of change sintroduced in many subsystems folders by this PR in every Phase workflow:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_0_X_2019-11-21-1100+e7fb7f/34293/dqm-histo-comparison-summary.html
I understand this is expected but I would like to have your confirmation

**Comment by cmantill** [comment]
@jfernan2 Yes, I think this is expected given that we are removing the old phase2 tracker customizations in this PR.

**Comment by kpedro88** [comment]
@cmantill thanks for this important development. A few questions and comments:
* To make sure I understand the scope, does this mean the hardcoded errors in https://github.com/cms-sw/cmssw/blob/9f9e834936b31e1c8fa77e358f8dd32de0f58420/RecoLocalTracker/SiPixelRecHits/src/PixelCPEGeneric.cc#L98-L134 will no longer be used?
* Currently, the T16 workflow uses the T15 global tag. Do all the T15 conditions actually work correctly for T16?
* The comparison differences are isolated to Phase 2 workflows and track-related quantities, and the extent of the changes seems minor:
log10 generaltrack pt
but it will still be good to see the full validation before approving the PR.

**Comment by mmusich** [comment]
@kpedro88,
re: 
> To make sure I understand the scope, does this mean the hardcoded errors in

These hardcoded errors were used in the generic CPE reco (not in the template one).
On the other hand this PR also provides GenError payloads and removes the customization in `RecoLocalTracker/SiPixelRecHits/python/PixelCPEGeneric_cfi.py` therefore yes, the harcoded errors won't be used anymore and could be removed (along with the configuration parameter that switches the on). I would propose this further cleanup for a follow-up PR.
re:
> Currently, the T16 workflow uses the T15 global tag. Do all the T15 conditions actually work correctly for T16?

Yes, the only difference between T15 and T16 is the skewed layout of TBPX. This at leading order does not affect CPE related quantities.

**Comment by mmusich** [comment]
@emiglior @pwittich FYI

File: CondFormats/SiPixelTransient/interface/SiPixelTemplate.h (hunk 1 above)
Comment (cmantill) [reply CondFormats/SiPixelTransient/interface/SiPixelTemplate.h]: @makortel do you have a specific memory check you would like to see? I am not sure if `SiPixelTemplateStore` is stored in stack. Maybe @mmusich is more familiar with this.

I ran the time and memory module comparison here:
**Before PR:**
`Summary for 10 events
Max VSIZ 6816.91 on evt 4 ; max RSS 3721.88 on evt 7
Time av 12.284 s/evt   max 26.0459 s on evt 4
CPU av 0 s/evt   max 0 s on evt -1
M1 Time av 11.3868 s/evt   max 26.0459 s on evt 4
M1 CPU av 0 s/evt   max 0 s on evt -1
M8 Time av 5.38321 s/evt   max 5.44807 s on evt 9`

**After PR:**
`Summary for 10 events
Max VSIZ 7344.61 on evt 6 ; max RSS 3759.03 on evt 7
Time av 16.4558 s/evt   max 33.244 s on evt 4
CPU av 0 s/evt   max 0 s on evt -1
M1 Time av 14.9319 s/evt   max 33.244 s on evt 4
M1 CPU av 0 s/evt   max 0 s on evt -1
M8 Time av 8.26217 s/evt   max 8.43981 s on evt 9`

I also ran the profile igprof, this shows the CPU use:
**After PR:**
https://cmantill.web.cern.ch/cmantill/cgi-bin//igprof-navigator/ig_phase2pr_Nov21_cpu/
or maybe more specifically:
https://cmantill.web.cern.ch/cmantill/cgi-bin//igprof-navigator/ig_phase2pr_Nov21_cpu/1283

And this shows the memory use (but I can't really tell anything from these logs so maybe you are more familiar navigating them):
**Before PR:**
https://cmantill.web.cern.ch/cmantill/cgi-bin//igprof-navigator/ig_phase2_nopr_Nov21
**After PR:**
https://cmantill.web.cern.ch/cmantill/cgi-bin//igprof-navigator/ig_phase2pr_Nov21

**Comment by kpedro88** [comment]
assign upgrade

File: CondFormats/SiPixelTransient/interface/SiPixelTemplate.h (hunk 1 above)
Comment (makortel) [reply CondFormats/SiPixelTransient/interface/SiPixelTemplate.h]: > do you have a specific memory check you would like to see? 

Mainly if I estimated the object size correctly or not.

> I am not sure if `SiPixelTemplateStore` is stored in stack. Maybe @mmusich is more familiar with this.

According to dxr
https://cmssdt.cern.ch/dxr/CMSSW/search?q=%2Btype-ref%3ASiPixelTemplateStore
the only place where the object is in stack is in
https://github.com/cms-sw/cmssw/blob/9f9e834936b31e1c8fa77e358f8dd32de0f58420/CondFormats/SiPixelTransient/src/SiPixelTemplate.cc#L165
which I suppose is fine as that code path is not run within CMSSW. And I see the corresponding piece in the code path run within CMSSW
https://github.com/cms-sw/cmssw/blob/9f9e834936b31e1c8fa77e358f8dd32de0f58420/CondFormats/SiPixelTransient/src/SiPixelTemplate.cc#L734-L735
was already change to heap allocation in https://github.com/cms-sw/cmssw/pull/22458.

Good, sorry for the noise.

**Comment by cmantill** [comment]
For reference, here are the slides from today's meeting where the consensus was to move forward with the PR: https://indico.cern.ch/event/838961/contributions/3648603/attachments/1950183/3237002/Template_Studies.pdf. Maybe @mmusich @tsusa or @emiglior can comment further.

**Comment by jfernan2** [comment]
+1
From DQM point of view

**Comment by kpedro88** [comment]
+upgrade

**Comment by christopheralanwest** [comment]
please test workflow 20007.0,20407.0,20807.0,21607.0,22007.0,22807.0,22834.0

**Comment by mmusich** [comment]
```
    if len(undefSet)>0: raise ValueError('Undefined workflows: '+', '.join(map(str,list(undefSet))))
ValueError: Undefined workflows: 22807.0, 21607.0, 20007.0, 20807.0, 20407.0, 22007.0
```
I am confused by this error. In the last IB (`CMSSW_11_0_X_2019-11-25-1100`)
```
runTheMatrix.py --what upgrade -nel 20007.0,20407.0,20807.0
,21607.0,22007.0,22807.0,22834.0
```
seems to run just fine.
Is the `--what upgrade` preventing to run the test in PRs?

**Comment by kpedro88** [comment]
@mmusich indeed, the PR tests do not know about the upgrade matrix.

If you want to forward a workflow from the upgrade matrix to the regular matrix, you can add it in one of these: https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/python/relval_2017.py, https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/python/relval_2026.py

**Comment by mmusich** [comment]
Thanks @kpedro88.
Perhaps @christopheralanwest can reformulate the tests by choosing one wf for each geometry from the ones already forwarded. We tested muon guns, but it does not necessarily needs to be it.

**Comment by kpedro88** [comment]
Yes, typically for phase 2 we just forward the ttbar workflows, since they produce enough particles to exercise most things. Since we have several scenarios in development, this limits the overall IB testing load.

**Comment by fabiocos** [comment]
@cmantill @mmusich the updated CondFormats affect also the digitization, so is this PR is expected to modify also that part? I see also an update of conditions.

BTW, @ggovi is this update ok?

**Comment by mmusich** [comment]
@fabiocos

> the updated CondFormats affect also the digitization, so is this PR is expected to modify also that part?

Mmh, does it?

> I see also an update of conditions.

The update in the SimLA should be purely nomenclature. The calibration constant un the two tags should coincide.

> BTW, @ggovi is this update ok?

The update is in the transient object, not in the one persisted in the DB, so there is no schema evolution.

**Comment by ggovi** [comment]
No Schema evolution issues.

**Comment by ggovi** [comment]
+1

**Comment by kpedro88** [comment]
please test workflow 20834.0,21634.0,22834.0

File: CondFormats/SiPixelTransient/interface/SiPixelGenError.h [hunk 2]
```diff
@@ -77,12 +77,12 @@ struct SiPixelGenErrorStore {  //!< template storage structure
   SiPixelGenErrorHeader head;
 #ifndef SI_PIXEL_TEMPLATE_USE_BOOST
   float cotbetaY[60];
-  float cotbetaX[5];
-  float cotalphaX[29];
+  float cotbetaX[60];
+  float cotalphaX[60];
   //!< 60 Barrel y templates spanning cluster lengths from 0px to +18px [28 entries for fpix]
   SiPixelGenErrorEntry enty[60];
   //!< 29 Barrel x templates spanning cluster lengths from -6px (-1.125Rad) to +6px (+1.125Rad) in each of 5 slices [3x29 for fpix]
```
[the remark this answers is not in the document]
Comment (cmantill) [reply CondFormats/SiPixelTransient/interface/SiPixelGenError.h]: @perrotta Ok, I implemented your suggested changes here: https://github.com/cms-sw/cmssw/pull/28448/commits/ef2998b90a784bd8ab539d5710acd514060eaa6d

**Comment by christopheralanwest** [comment]
please test workflow 20834.0,21634.0,22834.0

**Comment by kpedro88** [comment]
@mmusich can you confirm if this PR affects tracker digitization, or only affects downstream?

**Comment by mmusich** [comment]
@kpedro, 
it affects it in a maybe indirect way since the digitizer consumes a different LA tag, and maybe a different random number sequence is used, but other than statistical fluctuations I would not expect it to change the digitizer output. 
Just to understand, would it be a problem if digis were affected?

**Comment by kpedro88** [comment]
@mmusich if digis are affected, it changes the level of importance toward the HLT TDR, which will initially produce DIGI-RAW samples and then rereco them with 11_1 or 11_2.

To me, it doesn't look like there are any changes in the actual digis, but it would be good to confirm this ASAP.

**Comment by mmusich** [comment]
As I said no changes at the digi level are expected. The difference is meant purely at CPE (i.e. RECO) level.

**Comment by mmusich** [comment]
> which will initially produce DIGI-RAW samples and then rereco them with 11_1 or 11_2.

Said in another way if the preference is to postpone this development to 11.1 or 11.2 and this PR is merged in the meanwhile it won't break compatibility.

**Comment by kpedro88** [comment]
+upgrade

**Comment by jfernan2** [comment]
+1

**Comment by fabiocos** [comment]
@mmusich thanks for the clarification

**Comment by perrotta** [comment]
@cmantill , I made cpu timing measurements with and without this PR on top of the baseline (it was CMSSW_11_0_X_2019-11-27-2300) and I did not notice any significant increase in time at step3. I am happy of it, of course. But could you please tell me how did you make the timing measurements in https://github.com/cms-sw/cmssw/pull/28448#discussion_r349681037, which apparently reports about a sizeable processing time increase with this PR

**Comment by cmantill** [comment]
@perrotta I ran step3 with ` --customise Validation/Performance/TimeMemoryInfo.py`, in a cluster machine at LPC with nThreads=4, and then just used the log to compare the timing with `~/tools/getTimeMemSummary.sh log`, but if your tests don't show the same then is probably fine. I'd be happy to re-test with your commands if needed of course.

**Comment by perrotta** [comment]
> @perrotta I ran step3 with ` --customise Validation/Performance/TimeMemoryInfo.py`, in a cluster machine at LPC with nThreads=4, and then just used the log to compare the timing with `~/tools/getTimeMemSummary.sh log`, but if your tests don't show the same then is probably fine. I'd be happy to re-test with your commands if needed of course.

The overall job time could be affected by some normalization issue caused by other jobs running in parallel on the same machine. One should make sure that there are not concurrent jobs while running your step3.
I normally compare the cpu taken by the single modules using ```~/tools/timeDiffFromReport.sh```, as described in https://twiki.cern.ch/twiki/bin/viewauth/CMS/RecoIntegration#Compare_timing  (and you can even apply some normalization factor there, if it is needed to pick up which module modifies its time differently than all the other ones). By doing so I didn't notice relevant time differences in the affected modules: but I would be happy if you could cross-check it with your sample

**Comment by cmantill** [comment]
@perrotta Running again, and making sure other are no other concurrent jobs I actually see a slight decrease in time (might be random?)
*no PR* vs *with PR*
Job total:  5.9537 s/ev ==> 5.38926 s/ev

Here is the [diff log.](http://cmantill.web.cern.ch/cmantill/tmp/timediffpr28448.log) with `~/tools/timeDiffFromReport.sh`.

And running `~/tools/getTimeMemSummary.sh` I get:
*with PR*
```Summary for 10 events
Max VSIZ 6746.36 on evt 6 ; max RSS 3729.45 on evt 9
Time av 13.4345 s/evt   max 26.6682 s on evt 4
CPU av 0 s/evt   max 0 s on evt -1
M1 Time av 12.5571 s/evt   max 26.6682 s on evt 4
M1 CPU av 0 s/evt   max 0 s on evt -1
M8 Time av 6.11572 s/evt   max 6.24101 s on evt 10```
*no PR*
```Summary for 10 events
Max VSIZ 6687.9 on evt 3 ; max RSS 3703.52 on evt 7
Time av 15.9845 s/evt   max 35.3513 s on evt 4
CPU av 0 s/evt   max 0 s on evt -1
M1 Time av 14.2675 s/evt   max 35.3513 s on evt 4
M1 CPU av 0 s/evt   max 0 s on evt -1
M8 Time av 4.01985 s/evt   max 5.0875 s on evt 9
```

**Comment by perrotta** [comment]
Thank you @cmantill : this confirma my finding that timing is not affected by this PR

**Comment by perrotta** [comment]
+1
- CPE templates allowed for Phase2, with conditions taken from DB 
- Increased size of templates originates a manageable of the heap memory
- CPU processing time is not affected
- Jenkins tests pass and show minor changes in the Phase2 tracking related outputs compatible with the updates

**Comment by fabiocos** [comment]
@mmusich thanks for the confirmation, the specific objects modified in this PR do not seem to enter directly into the digitization (it is the SiPixelkTemplate2D object that enters for what I can see).

@christopheralanwest @tocheng @ggovi could you please check the latest update and comment or sign it?

**Comment by mmusich** [comment]
@fabiocos 

> not seem to enter directly into the digitization (it is the SiPixelTemplate2D object that enters for what I can see).

right. The 2D template objects are needed for the charge re-weighting which we'll use to inject the radiation damage simulation, but this will be provided later on as the phase-2 digitizer needs to be adjusted first to consume it. This will be one milestone development for the next cycle.

**Comment by fabiocos** [comment]
@ggovi the changes since your signature were merely technical could you pleas echeck and sign it again? In any case I will consider your previous signature as valid
@christopheralanwest @tuos @tlampen could you please have a final check and sign it or comment?

**Comment by ggovi** [comment]
+1

**Comment by tocheng** [comment]
+1

**Comment by fabiocos** [comment]
+1

**Pull request merged**